### ISSUE TITLE
explain, document, and test trial's compatibility requirements for other test tools

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2739]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#2739|
|Type|enhancement|
|Created|2007-07-20 14:02:24Z|
|Branch|https://github.com/twisted/twisted/tree/trial-compatibility-doc-2739|

The issue of trial needing to be compatible with other test runners is periodically raised, for example, in the comments on [#1870](https://github.com/twisted/twisted/issues/1870).  I believe Zope's test runner is the most popular of these, but I'm not sure.

It seems as though there is little information available for those of us who might want to make changes to trial as to what changes are acceptable and what changes are not, with respect to this requirement.

Are there tests within trial which verify this compatibility?  It looks like test_pyunitcompat is for running pyunit tests with trial, not trial tests with pyunit (or pyunit-like frameworks).  If they're somewhere else, where are they?  If not, can some be written?  With some dependency detection, perhaps, so we don't cause an earthquake on all the buildbots at once?

Is there documentation within trial which explains what the requirements of "other test runners" are?

What is the list of supported, external test runners?  Is this documented anywhere?

If all of these things exist, then this ticket could be resolved simply by making them easier to find.  Placing a link on any linked twistedmatrix.com page tends to push stuff to the top of google results pretty fast.  In searching around for a variety of phrases, [this file in Zope3 SVN was the only thing I came across](http://svn.zope.org/Zope3/trunk/trial.py?rev=41480&view=markup).

<details><summary>Searchable metadata</summary>

```
trac-id__2739 2739
type__enhancement enhancement
reporter__glyph glyph
priority__highest highest
milestone__ 
branch__branches_trial_compatibility_doc_2739 branches/trial-compatibility-doc-2739
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1184940144000000 1184940144000000
changetime__1195662333000000 1195662333000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Thanks for filing this ticket.

Since this is often given as a reason for some change to be made or not to be made to trial, resolving this is probably the highest priority trial ticket right now.  Any other potential bug fix or enhancement runs the risk of encountering a constraint which isn't documented or generally known.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|


Fair call.

Here is a rough draft:
 * Respect the interfaces defined in stdlib unittest.py
 * In particular, _TestCase.run(TestResult())_ should 'work' (i.e. complete synchronously, populate the result object etc)
 * If `trial foo.test_something.TestFoo.test_pass` works, then `TestFoo('test_pass').run(TestResult())` should also work, and vice versa.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

(In [21827]) Define `ITestCase` to be the interface that test frameworks expect of a test.

 * Author: jml
 * Reviewers: therve, exarkun
 * Fixes [#2898](https://github.com/twisted/twisted/issues/2898)

`ITestCase` is re-purposed to define what you can expect from a test object.
We use this interface to adapt pyunit tests to things usable in Trial.

In the process we also clarify what's needed for compatibility with doctest
and pyunit.

This branch also fixes [#1950](https://github.com/twisted/twisted/issues/1950).

Refs [#2739](https://github.com/twisted/twisted/issues/2739).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

(In [21903]) Branching to 'trial-compatibility-doc-2739'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @jml|
|-|-|

The URLs could use U{} to make them links in the output.  No need for re-review.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml set status to `closed`|
|-|-|

(In [21934]) Document Trial's compatibility requirements.

 * Author: jml
 * Reviewer: exarkun
 * Fixes [#2739](https://github.com/twisted/twisted/issues/2739)

Add information to the trial package docstring that starts to define the 
compatibility requirements for trial, giving examples.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
